### ISSUE TITLE
Ink v3 contains `@types` packages as dependencies

### ISSUE THREAD

**Issue by G-Rath** [vadimdemedes/ink#320]
Currently the `package.json` for v3 contains `@types/` packages as dependencies:

https://github.com/vadimdemedes/ink/blob/9ebfe314e865506ccb37b48723a27485346d950b/package.json#L41-L51

These should be listed as `devDependencies` :)

**Comment by SimenB** [comment]
the ones present in the `.d.ts` files should remain as deps, but I doubt all of these are (if any?)

**Comment by bartlangelaan** [comment]
I do think that it's better if they are listed as `*` instead of a version range, to make sure that only a single version per package is installed. That's also what the `@types/` packages themselves do.

**Comment by vadimdemedes** [comment]
> These should be listed as devDependencies :)

Could you elaborate why?

**Comment by G-Rath** [comment]
Sure! It's primarily for the same reasons that other dependencies are listed as `devDependencies`: they're not required by the library at runtime.

In theory none of these packages ever need to be listed as dependencies, but sometimes a library will use types from one or more of the packages in their typed API so those packages are sometimes included as `dependencies`. In that case, the definitions should usually be included using a looser version, depending on the type in question.

The other way you can do this is like what you've done with `@types/react`: mark the package as `optional` using `peerDependenciesMeta`.

Theres pros and cons to these two methods: the first (using `dependencies`) means that everyone will always get the `@types` packages that are required for TypeScript users to use your libraries types regardless of if they're using TypeScript or not; the second (using `peerDependenciesMeta`) means that the users typically have to manually include the `@types/` packages themselves if they're using TypeScript.

Looking over the codebase, I believe the only `@types` that should be in `dependencies` is `@types/react-reconciler`.

Constraint wise, typically `'*'` is recommended as mixing type versions can be messy, and it's usually easier to resolve problems with 1 package than with 2. The type packages are also not always 1:1 with the libraries they're typing in terms of major versions, so using `'*'` saves you from having to manage the overhead of doing i.e `24 | 28 | 29` as your constraint, nor will you have to make new releases updating those constraints because a new version of the `@types` was released & people are getting dependency warnings.

You should also always use `'*'` as the constraint for `@types/node` (if it's in `dependencies`, which it rarely needs to be), as it's the one dependency you can't actually control, and having two versions of it in your dependency tree can lead to some very confusing errors.

Hope that helps clarify things - let me know if you've got more questions! 😄

**Comment by vadimdemedes** [comment]
Yeah, that does make sense, thanks for detailed explanation!

> Looking over the codebase, I believe the only @types that should be in dependencies is @types/react-reconciler.

Maybe I'm missing something, but it doesn't look like I'm exporting anything from `@types/react-reconciler`?

**Comment by vadimdemedes** [comment]
The only thing I'm on the fence is about using `*` as semver for `@types/*` packages (except `@types/node`). Sounds like types could drift away from actual code in the package if `*` is used.

**Comment by SimenB** [comment]
FWIW, I think using versions instead of `*` makes sense since this project is written in TS, and relying on lockfiles or old `node_modules` to not get a type error seems like a code smell to me.

Regarding what to keep as deps, you can grep for `@types` in the build folder?

**Comment by G-Rath** [comment]
> it doesn't look like I'm exporting anything from @types/react-reconciler?

It's technically needed for this: https://github.com/vadimdemedes/ink/blob/6eaca9fae6570bc53e9663db4bd821c8f769422c/src/reconciler.ts#L42

This results in the following `reconciler.d.ts`:

```
import createReconciler from 'react-reconciler';
import type { TextNode, DOMElement } from './dom';
import './devtools';
declare const _default: createReconciler.Reconciler<DOMElement, DOMElement | TextNode, DOMElement, unknown>;
export default _default;
```

*But* looking deeper that doesn't seem to be part of the public API (aside from having a `.d.ts` as it's not imported by any of the other definition files, so if omitted it won't break.

I just did a small setup using the example in the README, removing `@types` (except for `react`), to confirm none of the `@types` packages are needed 🎉 

> The only thing I'm on the fence is about using * as semver for @types/* packages (except @types/node)
> FWIW, I think using versions instead of * makes sense since this project is written in TS

Just to be clear: I'm meaning using `*` for `@types` in `dependencies`, `@types` in `devDependencies` should definitely use standard version constraints, including `@types/node` (technically to be most accurate `@types/node` should be using the lowest major version allowed by your `engines` field).

`@types` in `dependencies` (excluding `@types/node`) can use version constraints too, but typically `'*'` is used for the reasons I mentioned before.

**Comment by vadimdemedes** [comment]
Landed in [3.0.0-7](https://github.com/vadimdemedes/ink/releases/tag/v3.0.0-7).
